### PR TITLE
🗺️ Implementation DiscoPOP Loss

### PR DESCRIPTION
**Pull request by fanconic**
# What does this PR do?
This PR implements the [DiscoPOP](https://huggingface.co/papers/2406.08414) loss. The DiscoPOP loss, also described in the paper as "Log Ratio Modulated Loss" (LRML), was discovered through an LLM discovery process to find optimization functions for offline preference optimization. The DiscoPOP loss outperformed (or performed competitively to) traditional offline preference optimization functions, such as DPO and Hinge loss, on different evaluation tasks (IMDb positive text generation, TL;DR text summarization, and Alpaca Eval 2.0).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

## Additional Notes
About tests, I have added tests to the best of my knowledge to make sure that this feature works. While test failures occurred in unrelated parts of the code, the ones I implemented passed successfully. Please let me know if something needs to be fixed.

### CODE REVIEW

File: docs/source/dpo_trainer.mdx [hunk 1]
```diff
@@ -150,6 +150,7 @@ The DPO algorithm supports several loss functions. The loss function can be set
...
+| `"discopop"`                           | The [DiscoPOP](https://huggingface.co/papers/2406.08414) paper uses LLMs to discover more efficient offline preference optimization losses. In the paper the proposed DiscoPOP loss (which is a log-ratio modulated loss) outperformed other optimization losses on different tasks (IMDb positive text generation, Reddit TLDR summarization, and Alpaca Eval 2.0). To use this discovered loss, set the `loss_type` value to `discopop` in the [`DPOConfig`].                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
```
Comment (qgallouedec) [inline docs/source/dpo_trainer.mdx:153]:
```suggestion
| `"discopop"`                           | The [DiscoPOP](https://huggingface.co/papers/2406.08414) paper uses LLMs to discover more efficient offline preference optimization losses. In the paper the proposed DiscoPOP loss (which is a log-ratio modulated loss) outperformed other optimization losses on different tasks (IMDb positive text generation, Reddit TLDR summarization, and Alpaca Eval 2.0).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
```

File: docs/source/dpo_trainer.mdx [hunk 2]
```diff
@@ -167,6 +168,10 @@ The [RPO](https://huggingface.co/papers/2404.19733) paper implements an iterativ
 
 The [WPO](https://huggingface.co/papers/2406.11827) paper adapts off-policy data to resemble on-policy data more closely by reweighting preference pairs according to their probability under the current policy. To use this method, set the `use_weighting` flag to `True` in the [`DPOConfig`].
 
+### DiscoPOP loss
+
+The [DiscoPOP](https://huggingface.co/papers/2406.08414) paper uses LLMs to discover more efficient offline preference optimization losses. In the paper the proposed DiscoPOP loss (which is a log-ratio modulated loss) outperformed other optimization losses on different tasks (IMDb positive text generation, Reddit TLDR summarization, and Alpaca Eval 2.0). To use this discovered loss, set the `loss_type` value to `discopop` in the [`DPOConfig`]. Additionally, you can change the `discopop_tau` value to change the shape of the DiscoPOP loss. However, the authors recommed the default value `discopop_tau=0.05`.
+
```
Comment (qgallouedec) [inline docs/source/dpo_trainer.mdx:174]:
```suggestion
```

Can you make the final remark about `discopop_tau` in the table instead?

File: trl/trainer/dpo_config.py [hunk 3]
```diff
@@ -65,6 +65,7 @@ class DPOConfig(TrainingArguments):
                 - `"aot_pair"`: AOT loss for unpaired datasets from the [AOT](https://huggingface.co/papers/2406.05882) paper.
                 - `"apo_zero"`: APO-zero loss from the [APO](https://huggingface.co/papers/2408.06266) paper.
                 - `"apo_down"`: APO-down loss from the [APO](https://huggingface.co/papers/2408.06266) paper.
+                - `"discopop"`: DiscoPOP (a.k.a Log-Ratio Modulated Loss, LRML) loss from the [DiscoPOP](https://huggingface.co/papers/2406.08414) paper.
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:68]: It's sorted by date

```suggestion
                - `"discopop"`: DiscoPOP (a.k.a Log-Ratio Modulated Loss, LRML) loss from the [DiscoPOP](https://huggingface.co/papers/2406.08414) paper.
                - `"apo_zero"`: APO-zero loss from the [APO](https://huggingface.co/papers/2408.06266) paper.
                - `"apo_down"`: APO-down loss from the [APO](https://huggingface.co/papers/2408.06266) paper.
```

Comment (qgallouedec) [reply trl/trainer/dpo_config.py:68]: Why not `"lrm"` instead by the way?

File: trl/trainer/dpo_config.py [hunk 4]
```diff
@@ -132,6 +133,9 @@ class DPOConfig(TrainingArguments):
             α parameter from the [RPO](https://huggingface.co/papers/2404.19733) paper (v3), which controls the
             weighting of the NLL term in the loss. If `None`, no weighting is applied and the loss is the same as the
             DPO loss. The paper recommends `rpo_alpha=1.0`.
+        discopop_tau (`float`, *optional*, defaults to `0.05`):
+            tau/temperature parameter from the [DiscoPOP](https://huggingface.co/papers/2406.08414) paper, which controls
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:137]:
```suggestion
            τ/temperature parameter from the [DiscoPOP](https://huggingface.co/papers/2406.08414) paper, which controls
```

File: trl/trainer/dpo_config.py [hunk 5]
```diff
@@ -176,6 +181,7 @@ class DPOConfig(TrainingArguments):
     ref_model_mixup_alpha: float = 0.9
     ref_model_sync_steps: int = 64
     rpo_alpha: Optional[float] = None
+    discopop_tau: Optional[float] = 0.05
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:184]:
```suggestion
    discopop_tau: float = 0.05
```

File: trl/trainer/dpo_trainer.py [hunk 6]
```diff
@@ -1019,11 +1019,26 @@ def dpo_loss(
             losses_chosen = F.sigmoid(self.beta * chosen_logratios)
             losses_rejected = 1 - F.sigmoid(self.beta * (chosen_logratios - rejected_logratios))
             losses = losses_chosen + losses_rejected
+            
+        elif self.loss_type == "discopop":
+            # Eqn (5) of the DiscoPOP paper (https://huggingface.co/papers/2406.08414)
+            # This loss was discovered with LLM discovery
+            pi_logratios = chosen_logps - rejected_logps
+            ref_logratios = ref_chosen_logps - ref_rejected_logps
+            logits = pi_logratios - ref_logratios
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1028]:
```suggestion
            logratios = chosen_logps - rejected_logps
            ref_logratios = ref_chosen_logps - ref_rejected_logps
            logits = logratios - ref_logratios
```

File: trl/trainer/dpo_trainer.py [hunk 7]
```diff
@@ -1019,11 +1019,26 @@ def dpo_loss(
...
+            exp_component = torch.exp(-logits)
+            # Blend between logistic and exponential component based on log ratio modulation
+            losses = logistic_component * (1 - log_ratio_modulation) + exp_component * log_ratio_modulation
+            return losses
 
         else:
             raise ValueError(
                 f"Unknown loss type: {self.loss_type}. Should be one of ['sigmoid', 'hinge', 'ipo', 'exo_pair', "
-                "'nca_pair', 'robust', 'bco_pair', 'sppo_hard', 'aot', 'aot_pair', 'apo_zero', 'apo_down']"
+                "'nca_pair', 'robust', 'bco_pair', 'sppo_hard', 'aot', 'aot_pair', 'apo_zero', 'apo_down', 'discopop']"
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1041]:
```suggestion
                "'nca_pair', 'robust', 'bco_pair', 'sppo_hard', 'aot', 'aot_pair', 'discopop', 'apo_zero', 'apo_down']"
```

File: trl/commands/scripts [hunk 8]
```diff
@@ -0,0 +1 @@
+/home/azureuser/caf83/trl/examples/scripts/
```
Comment (qgallouedec) [inline trl/commands/scripts:1]: please remove this file

**Comment by qgallouedec** [comment]
Thanks @fanconic! Do you have reference results to share?

File: trl/trainer/dpo_config.py (hunk 3 above)
Comment (fanconic) [reply trl/trainer/dpo_config.py:68]: I don't really have a clear answer to this. While `lrml` was the name proposed by the LLM during discovery, the authors agreed to name the best performing one `DiscoPOP`, which seemed to us like a catchy abbreviation for **D**iscovered **P**reference **Op**timization

File: trl/trainer/dpo_trainer.py [hunk 9]
```diff
@@ -1019,11 +1019,26 @@ def dpo_loss(
...
+            ref_logratios = ref_chosen_logps - ref_rejected_logps
+            logits = pi_logratios - ref_logratios
+            logits = logits * self.beta
+            # Modulate the mixing coefficient based on the log ratio magnitudes
+            log_ratio_modulation = torch.sigmoid(logits / self.args.discopop_tau)
+            logistic_component = -F.logsigmoid(logits)
+            exp_component = torch.exp(-logits)
+            # Blend between logistic and exponential component based on log ratio modulation
+            losses = logistic_component * (1 - log_ratio_modulation) + exp_component * log_ratio_modulation
+            return losses
```
[the remark this answers is not in the document]
Comment (fanconic) [reply trl/trainer/dpo_trainer.py:1036]: Good catch on this one. Apologies for it!

File: trl/commands/scripts (hunk 8 above)
Comment (fanconic) [reply trl/commands/scripts:1]: Sorry, this slipped in there when adding changed files. Good catch

File: trl/trainer/dpo_config.py [hunk 10]
```diff
@@ -150,6 +154,7 @@ class DPOConfig(TrainingArguments):
         "aot_pair",
         "apo_zero",
         "apo_down",
+        "discopop",
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:157]:
```suggestion
        "discopop",
        "apo_zero",
        "apo_down",
```

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
Thanks for contributing @fanconic 👊
